docs/source/expr-system-internals.rst

```diff
@@ -0,0 +1,206 @@
+
+Instead, expression classes use a dataclass-like interface defined by two attributes:
+
+* ``_parameters``: List of parameter names
+
+* ``_defaults``: Dictionary of default values for optional parameters
+
+Arguments passed to the constructor are stored in the operands attribute, with minimal input validation. Here's an example:
+
+.. code::
+
+    >>> class MyExpr(Expr):
+    >>>     _parameters = ["param1", "param2"]
+    >>>     _defaults = {"param2": None}
+
+
+    >>> class MyExpr(Expr):
+    >>>     _parameters = ["param1", "param2"]
+    >>>     _defaults = {"param2": None}
+
+    >>> expr = MyExpr(1, 2, 3)
+    >>> expr.param1
+    1
+
+    >>> expr.param2
+    2
+
+    >>> expr.operands
+    [1, 2, 3]
+
+Names and Tokens
+^^^^^^^^^^^^^^^^
+
+Every expression is uniquely identified by a name, composed of:
+
+* A prefix (typically the class name or a variant)
+
+* A token generated by hashing its operands via :func:`^dask.base.tokenize`
+
+This tokenization enables:
+
+* Deduplication of equivalent expressions in the graph
+
+* Detection of changes across optimization steps
+
+* Singleton enforcement for certain expression types
+
+Expressions that subclass :class:`^dask.dataframe._expr.SingletonExpr` (the default for most DataFrame expressions) are guaranteed to be unique by name. However, this tokenization system introduces several challenges:
+
+* Performance: Tokenization is slow due to recursive traversal and Dask’s dispatch mechanisms.
+
+* Determinism: Objects without a registered __dask_tokenize__ fallback to (cloud)pickling, which can be slow and non-deterministic.
+
+* Cross-interpreter behavior: Tokens are not consistent across interpreters or machines, complicating client-scheduler interactions.
+
+To address this, each expression computes and caches its name and token upon construction. These values are stored and serialized to ensure pickle roundtrip stability. The token is accessible via ``_determ_token`` or ``deterministic_token``.
+
+Caching and Singletons
+^^^^^^^^^^^^^^^^^^^^^^
+
+Despite efforts to keep expressions stateless, in practice many attributes are computed on demand and cached via ``functools.cached_property``. This defers computation but complicates reasoning about when and how state is evaluated.
+
+Cached properties are typically serialized with the expression, unless ``_pickle_functools_cache`` is set to False.
+
+To preserve cache values during repeated optimization (which recreates expressions), most classes subclass :class:`^dask.dataframe._expr.SingletonExpr`. This ensures that instances with the same name return a previously created, cached version. This makes expressions effectively immutable singletons — and must not be mutated in-place.
+
+Optimization Procedure
+----------------------
```

Comment:
This section seems more important to me than tokenization and caching.  I recommend moving it up.